sdk/iot/provisioning/samples/src/paho_iot_provisioning_example.c

```diff
@@ -0,0 +1,304 @@
+  }
+  else
+  {
+    printf("(missing) Please set the %s environment variable.\n", env_name);
+    return AZ_ERROR_ARG;
+  }
+
+  return AZ_OK;
+}
+
+static az_result read_configuration()
+{
+  az_span endpoint_span = AZ_SPAN_LITERAL_FROM_BUFFER(global_provisioning_endpoint);
+  AZ_RETURN_IF_FAILED(read_configuration_entry(
+      "Global Device Endpoint",
+      GLOBAL_PROVISIONING_ENDPOINT_ENV,
+      GLOBAL_PROVISIONING_ENDPOINT_DEFAULT,
+      false,
+      endpoint_span,
+      &endpoint_span));
+
+  az_span id_scope_span = AZ_SPAN_LITERAL_FROM_BUFFER(id_scope);
+  AZ_RETURN_IF_FAILED(read_configuration_entry(
+      "ID_Scope", ID_SCOPE_ENV, NULL, false, id_scope_span, &id_scope_span));
+
+  az_span registration_id_span = AZ_SPAN_LITERAL_FROM_BUFFER(registration_id);
+  AZ_RETURN_IF_FAILED(read_configuration_entry(
+      "Registration ID",
+      REGISTRATION_ID_ENV,
+      NULL,
+      false,
+      registration_id_span,
+      &registration_id_span));
+
+  az_span temp = AZ_SPAN_LITERAL_FROM_BUFFER(x509_cert_pem_file);
+  AZ_RETURN_IF_FAILED(read_configuration_entry(
+      "X509 Certificate PEM Store File", DEVICE_X509_CERT_PEM_FILE, NULL, false, temp, &temp));
+
+  AZ_RETURN_IF_FAILED(az_iot_provisioning_client_init(
+      &provisioning_client, endpoint_span, id_scope_span, registration_id_span, NULL));
+
+  return AZ_OK;
+}
+
+static int on_received(void* context, char* topicName, int topicLen, MQTTClient_message* message)
+{
+  UNUSED(context);
+  UNUSED(topicLen);
+
+  int i;
+  char* payloadptr;
+
+  printf("Message arrived\n");
+  printf("     topic: %s\n", topicName);
+  printf("   message: ");
+
+  payloadptr = message->payload;
+  for (i = 0; i < message->payloadlen; i++)
+  {
+    putchar(*payloadptr++);
+  }
+  putchar('\n');
+  MQTTClient_freeMessage(&message);
+  MQTTClient_free(topicName);
+  return 1;
+}
+
+static int connect()
+{
+  int rc;
+
+  MQTTClient_SSLOptions mqtt_ssl_options = MQTTClient_SSLOptions_initializer;
+  MQTTClient_connectOptions mqtt_connect_options = MQTTClient_connectOptions_initializer;
+
+  char username[128];
+  az_span username_span = AZ_SPAN_LITERAL_FROM_BUFFER(username);
+  if ((rc = az_iot_provisioning_client_user_name_get(
+           &provisioning_client, username_span, &username_span))
+      != AZ_OK)
+  {
+    printf("Failed to get MQTT username, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = az_span_append_uint8(username_span, '\0', &username_span)) != AZ_OK)
+  {
+    printf("Failed to get MQTT username, return code %d\n", rc);
+    return rc;
+  }
+
+  mqtt_connect_options.username = (char*)az_span_ptr(username_span);
+  mqtt_connect_options.password = NULL; // Using X509 Client Certificate authentication.
+
+  mqtt_ssl_options.keyStore = (char*)x509_cert_pem_file;
+  mqtt_connect_options.ssl = &mqtt_ssl_options;
+
+  if ((rc = MQTTClient_connect(mqtt_client, &mqtt_connect_options)) != MQTTCLIENT_SUCCESS)
+  {
```

Comment:
Can you check what is the scope of memory required by the option parameter.
I am just worried we gave usename that sits in stack.

Reply:
In general, Valgrind detection of overflows in stack and global variables is weak to non-existant. Arguably, Valgrind is the wrong tool for that job. If you are on one of supported platforms, building with -fmudflap and linking with -lmudflap will give you much better results for these kinds of errors.

---
In reply to: [406412124](https://github.com/Azure/azure-sdk-for-c/pull/547#discussion_r406412124) [](ancestors = 406412124)

Reply:
Here I was more concerned about stack memory.... I am just worried that if it start failing randomly... it will be harder to debug the issue.

---
In reply to: [406524774](https://github.com/Azure/azure-sdk-for-c/pull/547#discussion_r406524774) [](ancestors = 406524774,406412124)